proposals/NNNN-method-and-initializer-keypaths.md

```diff
@@ -0,0 +1,172 @@
+
+```swift
+struct Calculator {
+  func square(of number: Int) -> Int {
+    return number * number * multiplier
+  }
+
+  func cube(of number: Int) -> Int {
+    return number * number * number * multiplier
+  }
+
+  init(multiplier: Int) {
+    self.multiplier = multiplier
+  }
+
+  let multiplier: Int
+}
+
+// Key paths to Calculator methods
+let squareKeyPath = \Calculator.square
+let cubeKeyPath = \Calculator.cube
+```
+
+These key paths can then be invoked dynamically with a generic function:
+
+```swift
+func invoke<T, U>(object: T, keyPath: KeyPath<T, (U) -> U>, param: U) -> U {
+  return object[keyPath: keyPath](param)
+}
+
+let calc = Calculator(multiplier: 2)
+
+let squareResult = invoke(object: calc, keyPath: squareKeyPath, param: 3)
+let cubeResult = invoke(object: calc, keyPath: cubeKeyPath, param: 3)
+```
+
+Or used to dynamically create a new instance of Calculator:
+
+```swift
+let initializerKeyPath = \Calculator.Type.init(multiplier: 5)
+```
+
+This proposed feature homogenizes the treatment of member declarations by extending the expressive power of key paths to method and initializer members.
+
+## Detailed design
+
+Key path expressions can refer to instance methods, type methods and initializers, and imitate the syntax of non-key path member references.
+
+### Argument Application
+
+Key paths can reference methods in two forms:
+
+1. Without argument application: The key path represents the unapplied method signature.
+2. With argument application: The key path references the method with arguments already applied.
+
+Continuing our `Calculator` example, we can write either:
+
+```swift
+let squareWithoutArgs: KeyPath<Calculator, (Int) -> Int> = \Calculator.square
+let squareWithArgs: KeyPath<Calculator, Int> = \Calculator.square(of: 3)
+```
+
+If the member is a metatype (e.g., a static method, class method, initializer, or when referring to the type of an instance), you must explicitly include `.Type` in the key path root type.
+
+```swift
+struct Calculator {
+  static func add(_ a: Int, _ b: Int) -> Int {
+    return a + b
+  }
+}
+
+let addKeyPath: KeyPath<Calculator.Type, Int> = \Calculator.Type.add(4, 5)
+```
+
+Here, `addKeyPath` is a key path that references the add method of `Calculator` as a metatype member. The key path’s root type is `Calculator.Type`, and the value resolves to an applied instance method result type of`Int`. 
+
+### Overloads
+
+Keypaths to methods with the same base name and distinct argument labels can be disambiguated with explicit  argument labels:
+
+```swift
+struct Calculator {
+  var subtract: (Int, Int) -> Int { return { $0 + $1 } }
+  func subtract(this: Int) -> Int { this + this}
+  func subtract(that: Int) -> Int { that + that }
+}
+  
+let kp1 = \S.subtract // KeyPath<S, (Int, Int) -> Int
+let kp2 = \S.subtract(this:) // WritableKeyPath<S, (Int) -> Int>
+let kp3 = \S.subtract(that:) // WritableKeyPath<S, (Int) -> Int>
+let kp4 = \S.subtract(that: 1) // WritableKeyPath<S, Int>
```

Comment:
Just noticed: Would these be `WritableKeyPath` or just `KeyPath`?

Reply:
Yes, this should be just `KeyPath` - thank you for catching this!